core/shared/src/main/scala/fs2/internal/CompileScope.scala

```diff
@@ -88,11 +103,22 @@ private[internal] final class CompileScope[F[_]] private (val id: Token, private
    * If this scope is currently closed, then the child scope is opened on the first
    * open ancestor of this scope.
    */
-  def open: F[CompileScope[F]] = {
+  def open(interruptible: Option[(Effect[F], ExecutionContext)]): F[CompileScope[F]] = {
     F.flatMap(state.modify2 { s =>
       if (! s.open) (s, None)
       else {
-        val scope = new CompileScope[F](new Token(), Some(self))
+        val newScopeId = new Token
+        def iCtx = interruptible.map { case (effect, ec) =>
+          InterruptContext(
+            effect = effect
+            , ec = ec
+            , promise = Promise.unsafeCreate[F, Throwable](effect, ec)
+            , ref = Ref.unsafeCreate[F, (Option[Throwable], Boolean)]((None, false))
+            , interruptScopeId = newScopeId
+            , maxInterruptDepth = 256 // todo: want we somehow configure this ?
```

Comment:
Can we write a separate issue for configuring this value along with maxSteps and chunkSize in Algebra?

Reply:
are we ok with the side effects here (creating promise, ref, token)?
It's guarded by `flatMap`, but it's threading the line (however, it might be hard to write it otherwise)

Reply:
@SystemFw they are guarded by modify2, I think we are on the safe side here. The scope allocation is pretty costly now. It is also quite often, i.e. every `repeatPull` (that is used by `map` for example) closes and opens scope. So I would like to leave it as it is.